### PR TITLE
PLT-4167 Team Sidebar

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Team Sidebar

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-4167

#### API Reference PR
https://github.com/mattermost/mattermost-api-reference/pull/30

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by jasonblais** [comment]
Three minor notes, otherwise no issues. 

Will keep PM review label until Asaad's changes are in

----

1) Remove “Go to System Console” line from team selection page

image

2) If user has not joined any teams (and is therefore redirected to team selection page), change the "Back" button to "Logout", since 
 - "Back" doesn't do anything
 - User has no way to "Logout", if there's no teams he can join, or if there's no teams he wants to join

image

3) In platform meeting we discussed adding team sidebar behind a feature flag for now, at least until we've discussed the feature more closely with devs.

Hence, propose adding the setting `Show team sidebar if you've joined more than one team` to **Account Settings > Advanced > Preview pre-release features**, above the WebRTC setting

image

**Comment by it33** [comment]
#### Mention notification incorrectly shows when switching teams

Repro: 

1. User A is on Town Square on Team 2
2. User B is on Town Square on Team 2
3. User B does (at)all mention 
4. User A uses team sidebar to switch from Team 2 to Team 1

Observed: 

- In Step 4, User A gets a notification on Team 2 in the sidebar after switching to Team 1

BEFORE leaving Team 2

image

AFTER leaving Team 2

image

Expected: 

Option 1: 
- In Step 4, User A DOES NOT get a notification on Team 2 in the sidebar after switching to Team 1

Option 2: 
- In Step 3, User A gets a notification on Team 2 in the sidebar

**Comment by enahum** [comment]
does that repro every time?

I'll fix it :)

**Comment by it33** [comment]
Spoke with Elias and there's a couple small things in addition

1. Cache issue on notification switching from team to team
2. Sometimes an unread is displayed as a mention in team sidebar

Otherwise looks really nice!

**Comment by enahum** [comment]
@it33 @jasonblais bug fixes have been submitted, feel free to re-test

**Comment by jasonblais** [comment]
Sorry, this corner case is still reproducing:

1) If user has NOT joined any teams (and is therefore redirected to team selection page), change the "Back" button to "Logout", since 
 - "Back" doesn't do anything
 - User has no way to "Logout", if there's no teams he can join, or if there's no teams he wants to join

Steps to repro:
 - sign in as a Team or System Admin
 - for each team on the server, make sure **Team Settings > General > Allow any user with an account on this server to join this team** is set to "No"
 - leave all teams

Observed: "Back" button on top left corner
Expected: "Logout" button on top left corner

image

2) I found a new corner case:

 - sign in as a System Admin, leave all your teams
 - go to the System Console
 - click the three-dotted menu

Observed: Only option is to logout

image

Expected:  
 - If the System Admin has NOT joined any teams, display two options on the three-dotted dropdown menu: 

```
 "Team Selection" which redirects the user to the team selection page;
   ---- line separator ----
  "Logout"
```
 
 - If the System Admin HAS joined at least one team, do not change the existing behaviour. That is:
```
 Show list of teams the system admin has joined
   ---- line separator ----
  "Logout"
```

**Comment by enahum** [comment]
Cool I'll fix them both during the weekend

**Comment by enahum** [comment]
@jasonblais should be ok now

**Comment by jwilander** [comment]
Needs a rebase @enahum

**Comment by enahum** [comment]
done @jwilander

File: api/team.go [hunk 1]
```diff
@@ -31,6 +31,7 @@ func InitTeam() {
 	BaseRoutes.Teams.Handle("/all_team_listings", ApiUserRequired(GetAllTeamListings)).Methods("GET")
 	BaseRoutes.Teams.Handle("/get_invite_info", ApiAppHandler(getInviteInfo)).Methods("POST")
 	BaseRoutes.Teams.Handle("/find_team_by_name", ApiAppHandler(findTeamByName)).Methods("POST")
+	BaseRoutes.Teams.Handle("/members", ApiAppHandler(getMyTeamMembers)).Methods("GET")
```
Comment (enahum) [inline api/team.go]: will do

File: model/team_member.go [hunk 2]
```diff
@@ -16,6 +16,12 @@ type TeamMember struct {
 	DeleteAt int64  `json:"delete_at"`
 }
 
+type TeamMemberExtra struct {
+	TeamMember
+	MsgCount     int64 `json:"msg_count"`
+	MentionCount int64 `json:"mention_count"`
+}
```
Comment (enahum) [inline model/team_member.go]: sure, I knew something like this was going to happen, we can discuss this as a team and find a proper solution

File: webapp/components/team_sidebar/components/team_button.jsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2015 Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline webapp/components/team_sidebar/components/team_button.jsx]: will do

**Comment by enahum** [comment]
@jwilander ready for re-review

@jasonblais feel free to test it again

**Comment by jasonblais** [comment]
@enahum 

Issue 1:
1) Create two teams, and make each one public from "Team Settings" > "Allow any user with an account on this server to join this team" > "No"
2) Leave one of the teams
3) From Main Menu, choose "Join Another Team"

Observed:

At the top left window, link options reads "Logout"

image

Expected: At the top left window, link options reads "Back". 

NOTE: The behaviour is correct, as clicking the link brings the user back to the team they previously had active. Only the text is incorrect.

Issue 2:
1) Leave all your teams, get redirected to team selection page
2) Click "Logout" link at top left corner of the window

Observed: Nothing happens
Expected: User is logged out of the system

Issue 3:

The bug [it33 reported above](https://github.com/mattermost/platform/pull/4569#issuecomment-262992870) reproduces, but only partially. When you switch to another team in step 4, the mention jewel appears for a brief moment, then disappears.

**Comment by jasonblais** [comment]
+1 other than the issues above. No need to re-test bug fixes in a spinmint

File: model/client.go [hunk 4]
```diff
@@ -1735,6 +1735,30 @@ func (c *Client) GetTeamMembers(teamId string, offset int, limit int) (*Result,
...
+	} else {
+		defer closeBody(r)
+		return &Result{r.Header.Get(HEADER_REQUEST_ID),
+			r.Header.Get(HEADER_ETAG_SERVER), TeamMembersFromJson(r.Body)}, nil
+	}
+}
+
+// GetMyTeamMembers will return an array with TeamMemberUnread objects that contain the amount of
+// unread messages and mentions the current user has for the teams it belongs to. Must be authenticated.
+func (c *Client) GetMyTeamMembersUnread() (*Result, *AppError) {
```
Comment (jwilander) [inline model/client.go]: `GetMyTeamMembersUnread` is a little confusing, maybe `GetUnreadsForTeam` or `GetMyUnreadsForTeam`?

File: model/team_member.go [hunk 5]
```diff
@@ -16,6 +16,12 @@ type TeamMember struct {
 	DeleteAt int64  `json:"delete_at"`
 }
 
+type TeamMemberUnread struct {
```
Comment (jwilander) [inline model/team_member.go]: I think we should exclude the `Member` part here and just go with `TeamUnread` since really it doesn't have to have anything to do with the `TeamMember`, right?

File: webapp/actions/post_actions.jsx [hunk 6]
```diff
@@ -30,7 +30,9 @@ export function handleNewPost(post, msg) {
         }
     }
 
-    var websocketMessageProps = null;
+    AsyncClient.getMyTeamMembersUnread();
```
Comment (jwilander) [inline webapp/actions/post_actions.jsx]: Adding another request every time the client receives a post might affect performance negatively. Maybe only do this if the post is not for the current team? Or possibly we can just fake it on the client (might be harder than it seems since we have to take into account someone viewing different teams on different devices)

File: webapp/actions/global_actions.jsx [hunk 7]
```diff
@@ -534,3 +538,33 @@ export function emitBrowserFocus(focus) {
         focus
     });
 }
+
+export function redirectUserToDefaultTeam() {
+    const teams = TeamStore.getAll();
+    const teamMembers = TeamStore.getMyTeamMembers();
+    let teamId = BrowserStore.getGlobalItem('team');
```
Comment (hmhealey) [inline webapp/actions/global_actions.jsx]: Instead of storing this locally, this could be saved in prefs and be accessible across devices. I think that's how we do last channel viewed

File: webapp/actions/global_actions.jsx [hunk 8]
```diff
@@ -534,3 +538,33 @@ export function emitBrowserFocus(focus) {
...
+
+    if (!teamId && teamMembers.length > 0) {
+        let myTeams = [];
```
Comment (hmhealey) [inline webapp/actions/global_actions.jsx]: We could discuss this more, but I don't think it's worth it to put in the extra work to pick the user's first team alphabetically as opposed to just picking a team randomly on the first load, especially if we store this across all devices. By going alphabetically, there's an equal chance that we'll get the "correct" team as if we were to pick randomly. I think this was before you were with us, but this is the same reason that new users get 10 random users in their DM channels when they create their account instead of the first 10 alphabetically.

Alternatively, I don't think there's an issue with just showing the Select Team page if you don't have a "previous team" stored anywhere since it'll only happen once and then it'll be saved after that.

File: webapp/actions/global_actions.jsx [hunk 9]
```diff
@@ -534,3 +538,33 @@ export function emitBrowserFocus(focus) {
...
+        for (const index in teamMembers) {
+            if (teamMembers.hasOwnProperty(index)) {
+                const teamMember = teamMembers[index];
+                myTeams.push(Object.assign({
```
Comment (hmhealey) [inline webapp/actions/global_actions.jsx]: Even if we keep doing this alphabetically, this stuff with the `unread` and `mentions` count should be removed since it's not being used for anything else

File: webapp/actions/websocket_actions.jsx [hunk 10]
```diff
@@ -237,13 +245,18 @@ function handleLeaveTeamEvent(msg) {
 
         // if they are on the team being removed redirect them to the root
         if (TeamStore.getCurrentId() === msg.data.team_id) {
+            BrowserStore.removeGlobalItem('team');
```
Comment (hmhealey) [inline webapp/actions/websocket_actions.jsx]: What happens if you're removed from a team when you're offline? The next time you log in, will you get stuck in an error state since it's trying to switch to a team that you don't have access to?

File: webapp/components/sidebar_header_dropdown.jsx [hunk 11]
```diff
@@ -183,14 +177,14 @@ export default class SidebarHeaderDropdown extends React.Component {
...
-        var teamLink = '';
-        var inviteLink = '';
-        var manageLink = '';
-        var sysAdminLink = '';
-        var currentUser = this.props.currentUser;
-        var isAdmin = false;
-        var isSystemAdmin = false;
-        var teamSettings = null;
+        const currentUser = this.props.currentUser;
+        let teamLink = '';
```
Comment (hmhealey) [inline webapp/components/sidebar_header_dropdown.jsx]: 👍 to getting rid of `var`

File: webapp/components/signup/components/signup_ldap.jsx [hunk 12]
```diff
@@ -25,6 +26,8 @@ export default class SignupLdap extends React.Component {
     constructor(props) {
         super(props);
 
+        BrowserStore.removeGlobalItem('team');
```
Comment (hmhealey) [inline webapp/components/signup/components/signup_ldap.jsx]: Why is this here?

File: webapp/components/signup/signup_controller.jsx [hunk 13]
```diff
@@ -122,7 +122,9 @@ export default class SignupController extends React.Component {
             }
 
             if (userLoggedIn) {
-                browserHistory.push('/select_team');
+                GlobalActions.redirectUserToDefaultTeam();
+            } else {
+                BrowserStore.removeGlobalItem('team');
```
Comment (hmhealey) [inline webapp/components/signup/signup_controller.jsx]: Same as my last comment

File: webapp/actions/global_actions.jsx (hunk 7 above)
Comment (enahum) [inline webapp/actions/global_actions.jsx]: like it, will do

File: webapp/actions/global_actions.jsx (hunk 8 above)
Comment (enahum) [inline webapp/actions/global_actions.jsx]: this was following the PM spec's and the main idea is to get rid of the team selection page

File: webapp/actions/websocket_actions.jsx (hunk 10 above)
Comment (enahum) [inline webapp/actions/websocket_actions.jsx]: If we are going to move this to the preferences then this will have to change eventually

**Comment by enahum** [comment]
@jasonblais, @jwilander and @hmhealey ready for re-review

I renamed the functions to something more readable, used the preferences to store the last team and channel viewed.

Also enabled the team sidebar in mobile

File: api/post.go [hunk 14]
```diff
@@ -824,6 +824,8 @@ func sendNotifications(c *Context, post *model.Post, team *model.Team, channel *
 	message.Add("post", post.ToJson())
 	message.Add("channel_type", channel.Type)
 	message.Add("channel_display_name", channel.DisplayName)
+	message.Add("channel_name", channel.Name)
```
Comment (coreyhulen) [inline api/post.go]: is it `senderName` or `senderUsername`

File: api/team.go [hunk 15]
```diff
@@ -723,6 +730,28 @@ func getTeamByName(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getMyTeamMembers(c *Context, w http.ResponseWriter, r *http.Request) {
+	if result := <-Srv.Store.Team().GetTeamsForUser(c.Session.UserId); result.Err != nil {
```
Comment (coreyhulen) [inline api/team.go]: Isn't this already cached in the session stuff?

File: api/team.go [hunk 16]
```diff
@@ -801,6 +830,10 @@ func updateTeam(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	oldTeam.Sanitize()
 
+	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_UPDATE_TEAM, "", "", "", nil)
```
Comment (coreyhulen) [inline api/team.go]: Why a message for Inviting?  the user hasn't joined yet??? or am I missing something.

File: api/team.go [hunk 17]
```diff
@@ -723,6 +730,28 @@ func getTeamByName(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	} else {
+		data := result.Data.([]*model.TeamMember)
+		w.Write([]byte(model.TeamMembersToJson(data)))
+	}
+}
+
+func getMyTeamsUnread(c *Context, w http.ResponseWriter, r *http.Request) {
+	teamId := r.URL.Query().Get("id")
+
+	if result := <-Srv.Store.Team().GetTeamsUnreadForUser(teamId, c.Session.UserId); result.Err != nil {
```
Comment (coreyhulen) [inline api/team.go]: We should think about caching or etag, if this is going to be called all the time.

File: store/sql_preference_store.go [hunk 18]
```diff
@@ -326,3 +326,25 @@ func (s SqlPreferenceStore) Delete(userId, category, name string) StoreChannel {
 
 	return storeChannel
 }
+
+func (s SqlPreferenceStore) DeleteCategory(userId string, category string) StoreChannel {
```
Comment (coreyhulen) [inline store/sql_preference_store.go]: Unit test missing?

File: store/sql_team_store.go [hunk 19]
```diff
@@ -583,6 +583,32 @@ func (s SqlTeamStore) GetTeamsForUser(userId string) StoreChannel {
...
 }
 
+func (s SqlTeamStore) GetTeamsUnreadForUser(teamId, userId string) StoreChannel {
+	storeChannel := make(StoreChannel, 1)
+
+	go func() {
+		result := StoreResult{}
+
+		var members []*model.TeamUnread
+		_, err := s.GetReplica().Select(&members,
```
Comment (coreyhulen) [inline store/sql_team_store.go]: hmm...perf makes me nervous :)

File: api/post.go (hunk 14 above)
Comment (enahum) [inline api/post.go]: Oh that seems to be a bad merge, I'l fix this

File: api/team.go (hunk 15 above)
Comment (enahum) [inline api/team.go]: Totally forgot about that, I'll add the cache read here if it exists I'll return that

File: api/team.go (hunk 17 above)
Comment (enahum) [inline api/team.go]: totally agree

File: api/team.go (hunk 16 above)
Comment (enahum) [inline api/team.go]: This is to be able to update the UI when a team goes from close to open and vice-versa

File: store/sql_preference_store.go (hunk 18 above)
Comment (enahum) [inline store/sql_preference_store.go]: yup, I'll add it

File: store/sql_team_store.go (hunk 19 above)
Comment (enahum) [inline store/sql_team_store.go]: can you suggest a better way?

File: api/team.go (hunk 17 above)
Comment (enahum) [inline api/team.go]: So after discussing this, for now there's no need to add caching or etag to this as this function is only called when necessary and not every time

**Comment by enahum** [comment]
@coreyhulen issues have been addressed

**Comment by jasonblais** [comment]
Tested on Windows 10, Chrome
 - [each case described in spec](https://docs.google.com/document/d/11vpkjruX1dWtY5FF-hzGDNmYxO7Z_rsZVDQMOKxgJrc/edit#heading=h.kn51e2ntf0ps)
 - [each case found previously and reported in test doc](https://docs.google.com/document/d/159kGGff4YmDMObZn77d4W9Bg0J5F6XAyMx0VtOdk7n8/edit)
 - [each case reported previously on this thread](https://github.com/mattermost/platform/pull/4569)
 - [each case in "Multi-Team + DMs" release tests](https://docs.google.com/spreadsheets/d/1vnCBQ1Tb0h0L5vHe0qjsZ1BwAhqO1LkfXLHfgNz8rOM/edit#gid=1435727833)
 - randomly attempting to break the code, including different themes, mobile UX, notifications, themes per team

----

Couple of bugs:

##1

1) Make sure **System Console > Users and Teams > Enable Team Creation** is set to `True`
2) Click the three-dotted menu and choose "Create New Team"
3) On the team creation page, click "Back" from top left corner of the page

Observed: User redirected to team selection page (with no available teams)
image

Expected: User redirected back to the previous page (i.e. to previously active channel on previously active team)

##2

1) Go to Off -Topic channel of team 1
2) Switch to team 2
3) Switch back to team 1

Observed: Redirected to Town Square of team 1
Expected: Off-Topic opens in team 1

##3

Mention count on tab bar doesn't update until you've switched a team

Observed (after receiving (at)-mentions in Town Square):
image

Observed (after switching a team): mention count in tab bar is correct
image

##4

Leaving two teams in succession throws JavaScript errors

1) Join at least three teams
2) Click three-dotted menu and click "Leave Team"
3) Click three-dotted menu and click "Leave Team" (again)

Observed: Team is still visible. After trying to switch a channel, JS errors are thrown
image

##5

1) Make sure there is at least one team on a server
2) Create a new account
3) Sign in

Observed: User auto-joins the first team created on the server
Expected: User lands on team selection page since they haven't joined any team

##6

1) Have user1 and user2 in team A
2) Have user1 leave team A
3) Have user2 switch to team B, then back to team A

Observed (for user2): User1 appears as a member of the team and outside of the team in DM list

image


Expected (for user2): User1 only appears as a member outside of team A in DM list

NOTE: The expected behaviour happens after a refresh

----

A few other bugs found, but they occur on `master` and already have Jira tickets, e.g. https://mattermost.atlassian.net/browse/PLT-4833

----

Have messaged @asaadmahmood to take a look at the mobile UI

**Comment by jasonblais** [comment]
Also tested on the desktop app and found two bugs. Created a ticket for one of them (not caused by this PR) and will submit a ticket once this PR has merged (minor aesthetic UX defect)

File: api/channel.go [hunk 20]
```diff
@@ -843,17 +843,25 @@ func setLastViewedAt(c *Context, w http.ResponseWriter, r *http.Request) {
...
 	newLastViewedAt := int64(data["last_viewed_at"].(float64))
+	channelName := data["channel_name"].(string)
 
 	Srv.Store.Channel().SetLastViewedAt(id, c.Session.UserId, newLastViewedAt)
 
-	preference := model.Preference{
+	chanPref := model.Preference{
 		UserId:   c.Session.UserId,
-		Category: model.PREFERENCE_CATEGORY_LAST,
+		Category: c.TeamId,
```
Comment (hmhealey) [inline api/channel.go]: ~~Can you swap the Category and Name here so that the Category is 'last_channel' or whatever and the Name is the team id? That'll keep it consistent with how the team-specific themes are defined.~~

Also, have you thought about adding code to remove the old last channel preferences since they won't be used any more? It could be in a separate PR, but I think we might want to be careful about leaving unused preferences around since they'll increase the initial load slightly for every user

File: api/channel.go [hunk 21]
```diff
@@ -877,6 +885,8 @@ func updateLastViewedAt(c *Context, w http.ResponseWriter, r *http.Request) {
 		active = true
 	}
 
+	channelName := data["channel_name"].(string)
```
Comment (hmhealey) [inline api/channel.go]: This will cause the server to panic on bad input. It should be surrounded by an if statement like the one above for active. I'm 0/5 on if it should error out or provide a default though.

File: api/channel.go [hunk 22]
```diff
@@ -901,14 +911,21 @@ func updateLastViewedAt(c *Context, w http.ResponseWriter, r *http.Request) {
 		go clearPushNotification(c.Session.UserId, id)
 	}
 
-	preference := model.Preference{
+	chanPref := model.Preference{
 		UserId:   c.Session.UserId,
-		Category: model.PREFERENCE_CATEGORY_LAST,
+		Category: c.TeamId,
```
Comment (hmhealey) [inline api/channel.go]: Same comment as above about swapping the Category and Name

File: api/channel.go (hunk 21 above)
Comment (hmhealey) [inline api/channel.go]: Also, why is this done by channel name? I assume it's for the town square, but that'll break it if someone renames a channel

File: api/channel.go (hunk 20 above)
Comment (hmhealey) [inline api/channel.go]: Never mind, I saw your code below to delete the user's team-specific preferences when they leave a team. I actually like that much better. We don't need to change the existing themes code, but all future team-specific preferences should follow that pattern in the future

File: api/team.go [hunk 23]
```diff
@@ -358,6 +360,11 @@ func LeaveTeam(team *model.Team, user *model.User) *model.AppError {
 		return uua.Err
 	}
 
+	// delete the preferences that set the last channel used in the team and other team specific preferences
+	if result := <-Srv.Store.Preference().DeleteCategory(user.Id, team.Id); result.Err != nil {
```
Comment (hmhealey) [inline api/team.go]: Should this also be done when the user is removed from a team?

File: store/sql_team_store.go (hunk 19 above)
Comment (hmhealey) [inline store/sql_team_store.go]: Could we use a query that gets all of their ChannelMembers and then sums them up for each team in the Go code? That would remove a lot of strain from the DB and shouldn't be very expensive on the Go side

File: webapp/routes/route_team.jsx [hunk 24]
```diff
@@ -32,6 +33,7 @@ function doChannelChange(state, replace, callback) {
         channel = ChannelStore.getByName(state.params.channel);
 
         if (!channel) {
+            BrowserStore.setGlobalItem(TeamStore.getCurrentId(), state.params.channel);
```
Comment (hmhealey) [inline webapp/routes/route_team.jsx]: Is this still needed with the new preferences?

File: webapp/routes/route_team.jsx [hunk 25]
```diff
@@ -73,11 +75,13 @@ function preNeedsTeam(nextState, replace, callback) {
     GlobalActions.emitCloseRightHandSide();
 
     TeamStore.saveMyTeam(team);
+    BrowserStore.setGlobalItem('team', team.id);
```
Comment (hmhealey) [inline webapp/routes/route_team.jsx]: Same as last comment

File: api/channel.go (hunk 21 above)
Comment (enahum) [inline api/channel.go]: Oh yes you are right, but then I was doing it with the channel name and not the id cause you don't have the channels in the client for the teams you are not currently in, so that means that we are going to need to make a request if you don't have that channel somehow.

any other thoughts on this?

File: api/channel.go (hunk 22 above)
Comment (enahum) [inline api/channel.go]: nop, I need the last channel for every team that you are a member of and not just the last channel you saw

File: api/team.go (hunk 23 above)
Comment (enahum) [inline api/team.go]: hmm, isn't this function called for that case? Im almost sure it is

File: api/channel.go (hunk 22 above)
Comment (hmhealey) [inline api/channel.go]: Sorry, forgot to remove this comment when I realized why you did it this way :p

**Comment by enahum** [comment]
@jasonblais @hmhealey ready for one last review?

File: api/team.go (hunk 23 above)
Comment (hmhealey) [inline api/team.go]: For anyone else reading this, `LeaveTeam` is called when a user is removed from the team

**Comment by jasonblais** [comment]
I created improvement tickets for two cases I found, but not worth blocking the PR. 
 - https://mattermost.atlassian.net/browse/PLT-5038
 - https://mattermost.atlassian.net/browse/PLT-5037

Looking great!! Thanks everyone, excited to get this in soon!

A related PR from @asaadmahmood that should fix an alignment issue on mobile sidebar https://github.com/mattermost/platform/pull/4812

**Comment by enahum** [comment]
PR for mobile view already merged https://github.com/mattermost/platform/pull/4830 this can be merged at any time now
